toolchain/semantics/semantics_parse_tree_handler.cpp

```diff
@@ -130,27 +131,159 @@ auto SemanticsParseTreeHandler::PopScope() -> void {
+    -> SemanticsNodeId {
+  // TODO: This should attempt implicit conversions, but there's not enough
+  // implemented to do that right now.
+  if (from_type == SemanticsNodeId::BuiltinInvalidType ||
+      to_type == SemanticsNodeId::BuiltinInvalidType) {
+    return SemanticsNodeId::BuiltinInvalidType;
+  }
+  if (from_type == to_type) {
+    return from_type;
+  }
+  return SemanticsNodeId::Invalid;
+}
+
 auto SemanticsParseTreeHandler::TryTypeConversion(ParseTree::Node parse_node,
                                                   SemanticsNodeId lhs_id,
                                                   SemanticsNodeId rhs_id,
                                                   bool /*can_convert_lhs*/)
     -> SemanticsNodeId {
   auto lhs_type = semantics_->GetType(lhs_id);
   auto rhs_type = semantics_->GetType(rhs_id);
-  // TODO: This should attempt a type conversion, but there's not enough
-  // implemented to do that right now.
-  if (lhs_type != rhs_type) {
-    if (lhs_type != SemanticsNodeId::BuiltinInvalidType &&
-        rhs_type != SemanticsNodeId::BuiltinInvalidType) {
-      // TODO: This is a poor diagnostic, and should be expanded.
-      CARBON_DIAGNOSTIC(TypeMismatch, Error,
-                        "Type mismatch: lhs is {0}, rhs is {1}",
-                        SemanticsNodeId, SemanticsNodeId);
-      emitter_->Emit(parse_node, TypeMismatch, lhs_type, rhs_type);
+  // TODO: CanTypeConvert can be assumed to handle rhs conversions, and we'll
+  // either want to call it twice or refactor it to be aware of lhs conversions.
+  auto type = CanTypeConvert(rhs_type, lhs_type);
+  if (type.is_valid()) {
+    return type;
+  }
+  // TODO: This should use type names instead of nodes.
+  CARBON_DIAGNOSTIC(TypeMismatch, Error,
+                    "Type mismatch: lhs is {0}, rhs is {1}", SemanticsNodeId,
+                    SemanticsNodeId);
+  emitter_->Emit(parse_node, TypeMismatch, lhs_type, rhs_type);
+  return SemanticsNodeId::BuiltinInvalidType;
+}
+
+auto SemanticsParseTreeHandler::TryTypeConversionOnArgs(
+    ParseTree::Node arg_parse_node, SemanticsNodeBlockId /*arg_ir_id*/,
+    SemanticsNodeBlockId arg_refs_id, ParseTree::Node param_parse_node,
+    SemanticsNodeBlockId param_refs_id) -> bool {
+  CARBON_DIAGNOSTIC(NoMatchingCall, Error, "No matching callable was found.");
+
+  // If both arguments and parameters are empty, return quickly. Otherwise,
+  // we'll fetch both so that errors are consistent.
+  if (arg_refs_id == SemanticsNodeBlockId::Empty &&
+      param_refs_id == SemanticsNodeBlockId::Empty) {
+    return true;
+  }
+
+  auto arg_refs = semantics_->GetNodeBlock(arg_refs_id);
+  auto param_refs = semantics_->GetNodeBlock(param_refs_id);
+
+  // If sizes mismatch, fail early.
+  if (arg_refs.size() != param_refs.size()) {
+    CARBON_DIAGNOSTIC(CallArgCountMismatch, Note,
+                      "Received {0} arguments, but require {1} parameters.",
```

Comment:
"require N parameters" doesn't seem right to me. Maybe either
```suggestion
                      "Received {0} arguments, but require {1} arguments.",
```
or
```suggestion
                      "Received {0} arguments, but have {1} parameters.",
```
?